Tool Builder - Self Check Responses (FRONT-END)

jaybhagat opened the pull request:
## Notion ticket link
<!-- Please replace with your ticket's URL -->
[[Tool Builder] Self Check Response](https://www.notion.so/uwblueprintexecs/Task-Board-485f846ae00243599dddba2be9bcba82?p=eefac0f1739049ce92abbf45673b9f6b)


<!-- Give a quick summary of the implementation details, provide design justifications if necessary -->
## Implementation description
* Used array of breakpoints
* Each breakpoint has a component attached to it
* The component is created using the data from the breakpoint data in the array
* Creates the new breakpoint by setting the lower as the last breakpoint's upper + 1


<!-- What should the reviewer do to verify your changes? Describe expected results and include screenshots when appropriate -->
## Steps to test
1. Open http://localhost:3000/admin/dashboard/tools
2. Create or edit an existing tool
3. Go to Self Check tab
4. Click Response
5. Explore! You can try to add new breakpoints, change the screen size, set the min and max of the breakpoints, add questions, etc. Note that you will have to create questions for the self check that are either multiple choice or multiselect to be able to play around with this functionality.

NOTE: the back-end will be done in another PR.


<!-- Draw attention to the substantial parts of your PR or anything you'd like a second opinion on
For Frontend changes be sure to specify which routes/pages reviewers should check out!
 -->
## What should reviewers focus on?
* Error checking
* Functions to compute the selected questions titles... Should I use an array with the useState hook to keep track of the number for the questions that are added? ~~Would have to sort these though~~ Just realized that I still have to sort them (or add them in sorted order)..
* Any other suggestions to make the code better! :)
* Let me know if there's something I forgot with this ticket!


## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

ChinemeremChigbo reviewed (commented):
![image](https://user-images.githubusercontent.com/24593519/157777906-b54fd253-308f-4db0-ae4d-b6853eef5265.png)
Currently unable to set upper and lower for multiple choice like we talked about.

jennifer-lu reviewed (changes requested):
Great work Jay! The UI is looking good. I just have a few concerns after testing.

For the first breakpoint, the minimum value displayed is NaN, I think that displaying 0 would be friendlier to users:
<img width="972" alt="Screen Shot 2022-03-10 at 7 36 00 PM" src="https://user-images.githubusercontent.com/49915445/157779649-6b2947fd-9d1b-4aae-ade4-773216056d0d.png">

There is a bug in this edge case:
1. Make a Self Check with two multiple choice questions
2. For the Self Check Score select All Self Check Questions
3. Go back to the question tab and make one of the questions short answer
4. In the Self Check Score section of the Response tab, All Self Check Questions should be selected but it isn't:
<img width="1434" alt="Screen Shot 2022-03-10 at 7 45 58 PM" src="https://user-images.githubusercontent.com/49915445/157780254-e8d619ee-522d-4c09-a44e-d6abd1ce4221.png">

I can't seem to edit the maximum value of a break point after I set it.

jennifer-lu commented on src/components/toolSelfCheckResponse/index.tsx:
```diff
@@ -0,0 +1,145 @@
+                    textAlign="center"
+                    width="35%"
+                    display="inline-block"
+                    onBlur={(str) =>
```
Try changing `onBlur` to `onChange` to be able to edit the breakpoint score after unfocusing.

chamod-gamage reviewed (changes requested):
Looking great!
Just one change to the functionality:
The options available for scoring should correspond to those questions which have a numeric data type. This includes all sliders, and any of the other question types for which the numeric data type has been selected. (e.g. 
<img width="945" alt="image" src="https://user-images.githubusercontent.com/41309709/157781542-72af8cd3-0d4c-431e-8c2a-1aa19e554815.png">
)

jaybhagat commented:
Hey @ChinemeremChigbo @jennifer-lu @chamod-gamage, I fixed up the bugs and added the suggestions you guys had! Feel free to take a look again :) (Hopefully it'll be all good this time).

chamod-gamage reviewed (changes requested):
https://www.loom.com/share/188831d957144467b3f5c6c54a2696b6
Great work! Just a couple more quick fixes that I noted in this video^

Helpful SO link:
https://stackoverflow.com/questions/63224459/disable-scrolling-on-input-type-number-in-react

jennifer-lu reviewed (changes requested):
The fixes are all looking good, nice work!

I'm seeing a bug in a similar edge case as before. It's pretty nitpicky, but would be nice to fix:
1. Make a Self Check with two multiple choice questions (numeric)
2. For the Self Check Score select All Self Check Questions
3. Go back to the question tab and delete the first question
4. In the Self Check Score section of the Response tab, one question is selected twice:

![Screen Shot 2022-03-20 at 3 27 15 PM](https://user-images.githubusercontent.com/49915445/159179499-703baae2-8905-46d6-9137-37e520e6d74c.png)

5. Go back to the question tab an re-add the second numeric multiple choice
6. In the Self Check Score section of the Response tab, all questions are not selected but it shows they are:

![Screen Shot 2022-03-20 at 3 41 44 PM](https://user-images.githubusercontent.com/49915445/159179907-ba6c5dfe-9b22-43b5-adcc-bd0f6239f146.png)

Another bug comes up in this case:
1. Make a Self Check with one multiple choice question (numeric)
2. Only fill out values of the first two options, leaving the third empty
3. For the Self Check Score select All Self Check Questions
4. Minimum and maximum scores show NaN:

![Screen Shot 2022-03-20 at 3 51 03 PM](https://user-images.githubusercontent.com/49915445/159180745-5e654f4c-f25f-467f-9852-dd4c27bc7629.png)

I have some product-related questions that maybe @chamod-gamage can answer:
1. Should users be able to delete breakpoints?
2. The minimum score for multiple choices seems to be the lowest option value. I think that using 0 instead might be better since a user might not answer the question.

jennifer-lu reviewed (commented):
Please ignore this comment, just wanted to remove the request changes status. Oop unfortunately didn't work.

jaybhagat commented:
> I have a product-related question that maybe @chamod-gamage can also answer. The minimum score for multiple choices seems to be the lowest option value. Would using 0 instead be better since a user might not answer the question?

@jennifer-lu talked with @chamod-gamage and Jenna about this. Seems like all the questions are required so I just left it like it is.

Also added a delete button not part of the designs but discussed with PMs.

jennifer-lu reviewed (commented):
Looks amazing Jay! Thanks for adding the deletion feature and fixing the nitpicks. Everything is generally working well.

There seems to be a bug for the breakpoint keys:
1. Open the browser console
2. Set three breakpoints
3. Delete the second one
4. Add a third one and you'll see this in the console:
![Screen Shot 2022-03-27 at 8 02 53 PM](https://user-images.githubusercontent.com/49915445/160306978-1fbc481f-7231-4521-b262-281543a421c0.png)
5. Go to the questions tab and observe some unexpected behaviour (the delete breakpoint has reappeared):
![Screen Shot 2022-03-27 at 8 04 02 PM](https://user-images.githubusercontent.com/49915445/160307014-5efcbfd6-4c6f-4e53-8ec8-b17f3d92dd18.png)

jaybhagat commented:
@jennifer-lu @chamod-gamage just thought I should remind you guys about this PR! please review it whenever you get the time :)!

jennifer-lu reviewed (approved):
@jaybhagat Thanks for the reminder to review! This looks fantastic 🔥

I noticed a small nitpick:
1. Add a numeric multiple choice question with a max score of 5 (i.e. add option with corresponding value 5)
2. On the responses page select the question
3. On the question page make the question non-numeric
4. On the responses page the question is still selected and the maximum score is 5 despite the question being non-numeric:
<img width="968" alt="Screen Shot 2022-04-06 at 12 49 41 PM" src="https://user-images.githubusercontent.com/49915445/162027124-323d3ad6-3dd4-493a-baa5-0211f66abe65.png">
<img width="957" alt="Screen Shot 2022-04-06 at 12 52 28 PM" src="https://user-images.githubusercontent.com/49915445/162027002-868b3620-a2b6-4bd9-935f-940fb69822b9.png">

tawnyzhao commented on pages/admin/dashboard/toolBuilder.tsx:
```diff
@@ -16,16 +23,22 @@ import {
 
 export const getServerSideProps: GetServerSideProps = async ({ req, res }) => {
-    if (process.env.USE_ADMIN_LOGIN === "true") {
+    if (process.env.NODE_ENV == "production") {
```
This change should be reverted

tawnyzhao commented on pages/admin/dashboard/toolBuilder.tsx:
```diff
@@ -170,6 +207,203 @@ const ToolBuilder: Page = () => {
         getAllTools();
     }, []);
 
+    const getMinSum = (questions) =>
```
Try to move these functions outside of the component itself you don't need to use any closure values

tawnyzhao replied:
Any functions that need to interact with the component's state can stay inside the component body

tawnyzhao commented on pages/admin/dashboard/toolBuilder.tsx:
```diff
@@ -98,6 +121,20 @@ const ToolBuilder: Page = () => {
     );
+    const [selectedQuestions, setSelectedQuestions] = useState([]);
+    const [description, setDescription] = useState("");
+    const [breakpoints, setBreakpoints] = useState([
```
consider defining a type for breakpoint in this file and pass it to useState<type>

tawnyzhao reviewed (approved):
Looks really good so far! a few requests for changes

ChinemeremChigbo reviewed (approved):
Nicely done!
